### DISCUSSION TITLE
Reverse timeline?

### DISCUSSION THREAD

**Discussion by kallekn** [pachli/pachli-android#283] [Ideas]
It would be nice to have the option of reversing the timeline (displaying the newest posts in the bottom). This would be beneficial when reading the timeline in real chronological order, especially when reading long posts. This feature was available in some Twitter clients, and is currently available in Mona for iOS, but apparently in no Android app for Mastodon.

**Comment by nikclayton** [comment]
Thanks. I'm keeping this around as an idea, possible for the 3.x series.

**Comment by nikclayton** [comment]
After 2.7.0 is out I'll put a rough-and-ready implementation of this in Pachli Current. It'll probably take a couple of release cycles to iron out bugs, so I'll be relying on you for feedback (I don't know anyone else that has requested this feature...).

**Reply by kallekn** [reply]
Great, thank you! I am surprised that so many people don't seem to understand why this is preferable. We usually read most other things in that direction :-)

**Reply by nikclayton** [reply]
Yeah. The thing I found surprising when I looked in to this is how easily people flip between the two. For example, when viewing a thread the thread is presented in chronological order, oldest first, exactly opposite to the timeline. But people have no problem with that.

Some of it is going to be UI affordance -- there's no "pull up to refresh" gesture that people are familiar with (yet).

**Comment by nikclayton** [comment]
@kallekn There's a new Pachli Current rolling out now. If you open the preferences and scroll to the bottom you'll see a new "Lab experiements" item.

Tap that to get to this:

![image](https://github.com/user-attachments/assets/a406ccbc-c2d9-4034-922d-4ce7cf0896f4)

Toggle that on.

**Reply by kallekn** [reply]
Testing it now. Works beautifully. So much more intuitive to read in the right direction, feels like getting back my Plume Twitter client again! Pull up to refresh works fine, but there is no symbol showing that new posts are being checked, like there is when you pull down to refresh. So far, this is the only problem I've seen. Great work.

**Reply by kallekn** [reply]
Hm, actually, pull up to refresh doesn't seem to work. But "refresh" in the menu works as it should.

**Reply by politas** [reply]
Getting to that "Refresh" button is a little awkward. If there's no way to implement a "Pull up to refresh", could we get a button added at the bottom of the feed for refresh?

**Comment by nikclayton** [comment]
Reported issues with the feature so far (I'll keep this up to date as more are reported).

- No "pull up to refresh" equivalent
- Returning to the timeline can lose position (https://mastodonsweden.se/@kallekn/112972346059030009)

**Reply by kallekn** [reply]
Also, the profile pages don't display correctly. Or rather, you have to scroll all the way up to see the complete profile information. It would probably be best to keep the profile pages the way they are by default, with the newest posts on top, so one can see the complete profile information and the newest posts by a glance.

**Comment by zagor** [comment]
I just installed Pachli explicitly for this feature. It is so nice, thank you for adding it! ❤️

**Comment by politas** [comment]
I am finding my current position is not always being saved, so I need to scroll past a few days' worth of messages to get to new ones.

**Reply by kallekn** [reply]
Yes, this is a known bug. Hopefully it can be fixed some time in the future? In the meantime, I find that I mostly do not lose my position in the timeline if I avoid using the "back" button and the Pachli app icon for returning to the app after doing something else. Using the "recents" button on the home screen usually works. However, if I have posted something or edited a post, I more often than not lose my position.

What is your experience, when do you lose your position and when do you keep it?

**Reply by politas** [reply]
The longer the gap between when I last used Pachli, the more likely it is that I'll lose my position, it seems.

**Reply by zagor** [reply]
Do we have a proper bug ticket for this?

**Reply by zagor** [reply]
Until then: I've noticed that the app doesn't start in the wrong position, it actually knows where I was. For a fraction of a second after switching back, it shows my correct point in the timeline. Then it scrolls back X amount of time.
Is there anything we can do to help troubleshoot this?

**Reply by nikclayton** [reply]
Nothing to do to troubleshoot it, I think.

The problem is as follows (I think, untested):

1. When restoring the reading position, the underlying database library operates by row number, not the content of the row
2. So to restore position, Pachli finds the row number, and tells the database library to start there
3. If the reading position is reversed the counting at step 1 is counting from the wrong "end" of the timeline, so the number is off

Complicating this is the fact that the steps Pachli does to do this -- in theory -- shouldn't be necessary. The Android libraries that Pachli uses is supposed to manage saving and restoring the reading position automatically. That they're not -- and these hacks are necessary in the first place -- is an indication of a deeper problem that I haven't properly understood yet.

Right now progress on https://pachli.app/pachli/2024/08/02/harassment-controls.html is taking priority. I appreciate you all for trying out this experimental feature and reporting issues like this, but without anyone else stepping up to look into contributing a fix it's difficult to commit to when this feature will stop being an experiment and be properly supported.

**Reply by zagor** [reply]
I understand completely @nikclayton. I did not mean to add any pressure, only to offer help if possible. I am grateful that you have added this feature and I wish to place no untoward burden on your shoulders.

**Reply by kallekn** [reply]
Same. It is great to have this feature, even if it is experiemental and not perfect.

**Comment by politas** [comment]
I've raised a new issue on the main Mastodon github to have this feature implemented in the main web interface. Folks who like a natural reading order, please upvote it. https://github.com/mastodon/mastodon-android/issues/952

**Comment by zagor** [comment]
Version 2.10.1 seems to be better about remembering the last read position with natural timeline. Thank you! 👍

**Comment by chuckthenerd** [comment]
In 3.4.0, with reverse timeline enabled, consider applying sort method to the other tabs to the right of home icon (notifications, trending posts). Assumption is that sort would have 8h entry on top and 1h entry below (image ...)
![Screenshot_20260308_182035_Pachli](https://github.com/user-attachments/assets/35831cbd-0e03-472a-a45e-56f4f0ded727)

**Comment by Rhialto** [comment]
Thanks for having an option to read messages in chronological order! I'm still testing if Pachli properly remembers the place where I was in a timeline. I have 3 mastodon accounts and I switch between them. So far I have the feeling that it doesn't always keep the position, but it is a bit hard to keep track of mentally myself :-)
